src/Common/rename.cpp

```diff
@@ -0,0 +1,122 @@
+#include <Common/rename.h>
+#include <Common/Exception.h>
+#include <Poco/File.h>
+
+#if defined(linux) || defined(__linux) || defined(__linux__)
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/fs.h>
+#include <sys/utsname.h>
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int ATOMIC_RENAME_FAIL;
+    extern const int SYSTEM_ERROR;
+    extern const int UNSUPPORTED_METHOD;
+    extern const int FILE_ALREADY_EXISTS;
+}
+
+static bool supportsRenameat2Impl()
+{
+#if defined(__NR_renameat2)
+    /// renameat2 is available in linux since 3.15
+    struct utsname sysinfo;
+    if (uname(&sysinfo))
+        return false;
+    char * point = nullptr;
+    auto v_major = strtol(sysinfo.release, &point, 10);
+
+    errno = 0;
+    if (errno || *point != '.' || v_major < 3)
+        return false;
+    if (3 < v_major)
+        return true;
+
+    errno = 0;
+    auto v_minor = strtol(point + 1, nullptr, 10);
+    return !errno && 15 <= v_minor;
+#else
+    return false;
+#endif
+}
+
+#if defined(__NR_renameat2)
+
+static void renameat2(const std::string & old_path, const std::string & new_path, int flags)
+{
+    if (old_path.empty() || new_path.empty())
```

Comment:
Please, use more detailed exception message.
Like `Cannot rename old_path to new_path because one of them is empty` or `Cannot rename old_path to new_path because one of them is absolute`.